src/core/net/dnssd_server.cpp

```diff
@@ -113,166 +113,170 @@ void Server::HandleUdpReceive(Message &aMessage, const Ip6::MessageInfo &aMessag
-                                      aName + nameComponentsInfo.mDomainOffset);
+    // Answer the questions with additional RRs if required
+    VerifyOrExit(resolveAdditional != kResolveNone);
+
+    readOffset = sizeof(Header);
+    for (uint16_t i = 0; i < aRequestHeader.GetQuestionCount(); i++)
+    {
+        IgnoreError(Dns::Name::ReadName(aMessage, readOffset, name, sizeof(name)));
+        IgnoreError(aMessage.Read(readOffset, question));
+        readOffset += sizeof(question);
+
+        VerifyOrExit(Header::kResponseServerFailure !=
+                         ResolveQuestion(name, question, responseHeader, aResponse, compressInfo, resolveAdditional),
+                     response = Header::kResponseServerFailure);
+
+        otLogInfoDns("[server] ADDITIONAL: TRANSACTION=0x%04x, QUESTION=[%s %d %d], RCODE=%d",
+                     aRequestHeader.GetMessageId(), name, question.GetClass(), question.GetType(), response);
+    }
 
 exit:
-    return error;
+    response = (error == OT_ERROR_NONE) ? response : Header::Response::kResponseServerFailure;
+
+    if (response == Header::Response::kResponseServerFailure)
+    {
+        otLogWarnDns("[server] failed to handle DNS query due to server failure");
+        responseHeader.SetQuestionCount(0);
+        responseHeader.SetAnswerCount(0);
+        responseHeader.SetAdditionalRecordCount(0);
+        IgnoreError(aResponse.SetLength(sizeof(Header)));
+    }
+
+    responseHeader.SetResponseCode(response);
+    aResponse.Write(0, responseHeader);
 }
 
-otError Server::ResolveQuestion(const char *      aName,
-                                const Question &  aQuestion,
-                                Header &          aResponseHeader,
-                                Message &         aResponseMessage,
-                                NameCompressInfo &aCompressInfo)
+Header::Response Server::ResolveQuestion(const char *      aName,
+                                         const Question &  aQuestion,
+                                         Header &          aResponseHeader,
+                                         Message &         aResponseMessage,
+                                         NameCompressInfo &aCompressInfo,
+                                         uint8_t           aResolveKind)
 {
     OT_UNUSED_VARIABLE(aName);
     OT_UNUSED_VARIABLE(aQuestion);
     OT_UNUSED_VARIABLE(aResponseHeader);
     OT_UNUSED_VARIABLE(aResponseMessage);
     OT_UNUSED_VARIABLE(aCompressInfo);
 
-    otError error = OT_ERROR_NONE;
+    Header::Response response = Header::kResponseNameError;
 
 #if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE
-    SuccessOrExit(error = ResolveQuestionBySrp(aName, aQuestion, aResponseHeader, aResponseMessage,
-                                               /* aAdditional */ false, aCompressInfo));
+    response = ResolveQuestionBySrp(aName, aQuestion, aResponseHeader, aResponseMessage, aResolveKind, aCompressInfo);
```

Comment:
nit: the order of `aResolveKind` and `aCompressInfo` are swapped relative to `Server::ResolveQuestion()`. Is this intentional?

Reply:
No, it's not intentional. I will make them uniform.